internal/brokers/broker.go

```diff
@@ -369,7 +369,11 @@ func validateUserInfoAndGenerateIDs(selectedUsername string, uInfo userInfo) (us
 	if uInfo.Name == "" {
 		return users.UserInfo{}, fmt.Errorf("empty username")
 	}
-	if uInfo.Name != selectedUsername {
+	// Ensure that the name of the user which authenticated matches
+	// the name which the user tries to log in as.
+	// Microsoft Entra user names are case insensitive, so we ignore
+	// the case when comparing the names.
+	if strings.ToLower(uInfo.Name) != strings.ToLower(selectedUsername) {
```

Comment:
The linter is already complaining, but I'll add a comment anyway: you should use the EqualFold func instead ([look here](https://github.com/ubuntu/authd-oidc-brokers/blob/0341cf9017e6c6f8aaca791c1b409f6a5ca02735/internal/broker/broker.go#L709)).

Reply:
I wonder if we should lowercase the username completely in authd... Wdyt?

cc. @didrocks

Reply:
sssd doesn’t do that AFAIK as it seems the first authentication case is always kept (in particular in home user creation). I would go with the comparison check only, which I thought was the first implementation we had here.

Reply:
ok, my memory isn’t that bad, we are doing it, but in the oidc-brokers* only: https://github.com/ubuntu/authd-oidc-brokers/blob/main/internal/broker/broker.go#L709

Reply:
I didn't see that you already commented on my draft PR ;) I saw the linter warning and decided that we don't want to use `strings.EqualFold`, because we don't know if Microsoft Entra does other case folding than simple lowercasing.

Reply:
Yeah... I didn't realize it was a draft until I commented :sweat_smile:...

The [documentation](https://pkg.go.dev/strings#EqualFold) states that it uses a simple form of case-folding, so we shouldn't have any problems with UTF-8 shenanigans (see [their example](https://pkg.go.dev/strings#example-EqualFold))

Reply:
Also, if a broker supports utf-8 usernames it should probably adjust them at broker level, since unix user names can't be utf-8 anyways.

Reply:
What I want to avoid is that another user on the same Entra instance can log in as another user because `strings.EqualFold` considers different Entra user names equal. For example, `strings.EqualFold` considers the letters "İ" and "I" both equal to "i" (see https://go.dev/play/p/HV2FKBBX-Xs). Si if Entra supports usernames with the letter "İ" (I don't know if it does, I'm just trying to illustrate the problem), then there could be two Entra users "Incognito@example.onmicrosoft.com" and "İncognito@example.onmicrosoft.com" which `strings.EqualFold` both considers equal. So if the owner of the "İncognito" account logs in as "Incognito" but authenticates as "İncognito", authd would log them in as the wrong user. Does that make sense?

Reply:
> Also, if a broker supports utf-8 usernames it should probably adjust them at broker level, since unix user names can't be utf-8 anyways.

I'm able to create a local user İncognito like this:
```
sudo adduser --allow-all-names İncognito
```

Reply:
Oh, I see, well...

```
The  least  restrictive  policy, available by using the --allow-all-names option, simply makes
 the same checks as useradd(8): cannot start with a dash, plus sign, or tilde; and cannot  con‐
 tain a colon, comma, slash, or whitespace.                                                       
                                                                                                  
 This option can be used to create confusing or misleading names; use it with caution.
```

Those are technically working, but not following the original `IEEE 1003.1-2001`. Which is still what we mostly support anyways. For example, both [ubuntu installer](https://github.com/canonical/ubuntu-desktop-provision/blob/main/provd/internal/services/user/user.go#L25-L26) and the desktop initial setup, follow still that rule, and I feel that not doing that could break the assumptions of some tools around.

So... while I agree that we should potentially consider that case, I feel that `authd` should (at least for now), just disallow the users that don't match the `"^[a-z_][a-z0-9_-]*$"` regex, as we do in most of places so far.

Reply:
Whatever decision is taken (I think @adombeck rationale is correct here), please ensure the authd-oidc-brokers aligns here too (see the previous link I reported).

Reply:
I don't understand why we have this check twice, once in the broker and once in authd. I think the broker is the correct place for it, because it's specific to the provider, and so is the username check (e.g. usernames of other providers might be case-sensitive). How about we drop this check here and trust that the broker returns a valid username?

Reply:
That was my initial reaction when I saw this PR.

Let’s see what @denisonbarbosa thinks, but I agree with the rationale to put a little bit more trust in the broker. This gives flexibility too for the broker, at the cost of them potentially introducing security vulnerability by not doing this check.

Reply:
> How about we drop this check here and trust that the broker returns a valid username?

done